Refactor ABFS tests to be more extensible

jirassimok opened the pull request:
This is the refactoring and additional tests added in preparation for #5052 supporting a new authentication method for ABFS.

findepi commented:
> This is currently just #3044 rebased onto master.

@jirassimok @electrum  any reason not to rebase the original one?

jirassimok commented:
@losipiuk This is all but the last commit from #5052, plus tests for `PrestoAzureConfigurationInitializer`.

jirassimok commented:
@findepi Try updating #5122 again, I fixed the mistake that was causing the errors (I had rebased d1ac86c improperly onto abdc5f2 and was restarting the wrong services in the container).

electrum commented on presto-hive/src/main/java/io/prestosql/plugin/hive/azure/PrestoAzureConfigurationInitializer.java:
```diff
@@ -36,39 +36,37 @@
     private final Optional<String> abfsAccessKey;
     private final Optional<String> abfsStorageAccount;
 
+    private static Optional<String> dropEmpty(Optional<String> optional)
```
Private helper methods should go after public methods

electrum commented on presto-hive/src/main/java/io/prestosql/plugin/hive/azure/PrestoAzureConfigurationInitializer.java:
```diff
@@ -36,39 +36,37 @@
-                    wasbStorageAccount.isPresent() && !wasbStorageAccount.get().isEmpty(),
-                    "hive.azure.wasb-access-key is set, but hive.azure.wasb-storage-account is not");
+                    wasbAccessKey.isPresent() && wasbStorageAccount.isPresent(),
+                    "If WASB storage account or access key is set, both must be set.");
```
Don't end messages in a period, unless there are multiple sentences

electrum commented on presto-hive-hadoop2/bin/common.sh:
```diff
@@ -142,3 +142,54 @@ function get_hive_major_version() {
+    done
+    exec_in_hadoop_master_container bash -c '
+        sed "${@:2}" "/docker/files/$1" > /etc/hadoop/conf/core-site.xml' \
+                                    bash "$template" "${args[@]}"
```
This indentation seems off. Probably better not to wrap here

jirassimok replied:
I've fixed the indentation, but the command being given to `bash` feels harder to follow when the line isn't wrapped.

electrum commented on presto-hive-hadoop2/bin/run_hive_abfs_tests.sh:
```diff
@@ -12,43 +12,16 @@ cleanup_hadoop_docker_containers
-    -e "s|%ABFS_ACCESS_KEY%|${ABFS_ACCESS_KEY}|g" \
-    -e "s|%ABFS_ACCOUNT%|${ABFS_ACCOUNT}|g" \
-    /etc/hadoop/conf/core-site.xml
+deploy_core_site_xml core-site.xml.abfs-template ABFS_ACCESS_KEY ABFS_ACCOUNT
```
Wrap the environment variables to be consistent

electrum commented on presto-hive-hadoop2/bin/run_hive_abfs_tests.sh:
```diff
@@ -12,43 +12,16 @@ cleanup_hadoop_docker_containers
-test_root="abfs://${ABFS_CONTAINER}@${ABFS_ACCOUNT}.dfs.core.windows.net/${test_directory}"
 
 # insert Azure credentials
 # TODO replace core-site.xml.abfs-template with apply-site-xml-override.sh
```
I'm not even sure what this comment means. Let's remove it everywhere.

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/azure/TestPrestoAzureConfigurationInitializer.java:
```diff
@@ -0,0 +1,121 @@
+     * {@link HiveAzureConfig} setter method. Use via {@link #toConfig}.
+     */
+    private static final Map<String, Method> PROPERTY_SETTERS =
+            ConfigurationMetadata.getValidConfigurationMetadata(HiveAzureConfig.class)
```
Static import `getValidConfigurationMetadata` and format like
```java
getValidConfigurationMetadata(HiveAzureConfig.class).getAttributes().values().stream()
        .map(AttributeMetadata::getInjectionPoint)
        .collect(toMap(InjectionPointMetaData::getProperty, InjectionPointMetaData::getSetter));

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/azure/TestPrestoAzureConfigurationInitializer.java:
```diff
@@ -0,0 +1,121 @@
+                            InjectionPointMetaData::getProperty,
+                            InjectionPointMetaData::getSetter));
+
+    private static final Set<Set<String>> propertyGroups = Set.of(
```
This is constant so it should be uppercase

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/azure/TestPrestoAzureConfigurationInitializer.java:
```diff
@@ -0,0 +1,121 @@
+    @DataProvider(parallel = true)
+    public Iterator<Object[]> propertyGroups()
+    {
+        return propertyGroups.stream().map(p -> new Object[]{p}).iterator();
```
Name it `property`. There should a space between `[]` and `{`

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/azure/TestPrestoAzureConfigurationInitializer.java:
```diff
@@ -0,0 +1,121 @@
+    @Test(dataProvider = "propertyGroups")
+    public void testPropertyGroups(Set<String> properties)
+    {
+        var config = toConfig(properties);
```
Avoid using `var` here since the type is not clear.

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/azure/TestPrestoAzureConfigurationInitializer.java:
```diff
@@ -0,0 +1,121 @@
+    public void testPropertyGroups(Set<String> properties)
+    {
+        var config = toConfig(properties);
+        try {
```
No need for try/catch since the test will fail anyway if it throws

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/azure/TestPrestoAzureConfigurationInitializer.java:
```diff
@@ -0,0 +1,121 @@
+import static org.testng.Assert.assertThrows;
+import static org.testng.Assert.fail;
+
+public class TestPrestoAzureConfigurationInitializer
```
While clever, this whole class is hard to read, hard to tell what it's doing, and hard to tell that it's correct. Test code should be a simple and clear as possible. Avoiding redundancy is not a goal. The fact that we're basically writing a framework to reflectively call methods says that this is too complicated.

jirassimok replied:
Yes, I overdid this a bit.

~~My first thought for how to simplify is to replace `PROPERTY_SETTERS` with a literal map listing the actual setters. That eliminates the reflection, but leaves most of the other complexity and still has helpful test output (including property names).~~

~~My second thought is to remove the property names entirely and replace the whole thing with a setup that has tests that look like this:~~
```
@Test
public void testAbfsAccessKey() {
    testPropertyGroup(HiveAzureConfig::setAbfsAccessKey, HiveAzureConfig::setAbfsStorageAccount);
}
```
~~The class is much easier to follow.~~

On second thought, I'll just go with the second idea. These properties aren't complicated enough that the tests really need to output exactly which properties are set.

electrum reviewed (commented):
A few minor comments. Overall looks good

losipiuk commented:
LGTM % David's comments

losipiuk commented:
Automation validated via https://github.com/prestosql/presto/pull/5122
